README.md

```diff
@@ -1 +1,189 @@
+>
+> — [RFC8259](https://www.rfc-editor.org/rfc/rfc8259#page-10)
+
+``` json
+{
+  "role": "user",  // Parsed by an IPLD implementation
+  "role": "admin", // Malicious duplicate field, omitted by the IPLD parser, accpeted by the browser
+  "links": [
+    {"/": "bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"},
+    {"/": "bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"},
+    {"/": "bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"}
+  ],
+  "sig": "8ufaS9w3CGN8cbQTUSoL1i7eaKiWLSXsD2LbZVmvM9zF"
+}
+```
+
+In the above example, the canonicalization step MAY lead to the signature validating, but the client parsing the `role: "admin"` field instead.
+
+# 2 Safety
+
+Data that has already been parsed to an in-memory IPLD representation can be canonically encoded to a trivially: it has already been through a [parser / validator](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).
+
+Data purporting to conform to an IPLD encoding (such as [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/)) MUST be validated prior to signature verification. This can be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signalled if it does not match. 
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec](https://ipld.io/specs/codecs/dag-json/spec/)
+
+As it is critical for signatures guard against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x00`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+A varsig is a bytestring that includes the following information:
+
+| Segment Name              | Type     | Description                                     | Required |
+|---------------------------|----------|-------------------------------------------------|----------|
+| Varsig Prefix             | `0xd000` | The multicodec varsig prefix                    | Yes      |
+| Public Key Prefix         | `Varint` | The multicodec prefix for the public key type   | Yes      |
+| Hash Prefix               | `Varint` | The multicodec prefix for the hash algorithm    | Yes      |
+| Hash Length               | `Varint` | The hash length                                 | Yes      |
+| Content Multicodec Prefix | `Varint` | The IPLD encoding uses to canonicalize the data | Yes      |
```

Comment:
I think this one needs to go up because length usually goes right before the thing it's denotes length of. Maybe it can be before Hash Prefix ?

Also per @mikeal's suggestion this should be a `Multiformat` not `Varint` this would allow specific ones to include more info in them.

Reply:
Done 👍